packages/macros/src/logging.rs

```diff
@@ -0,0 +1,195 @@
+        Ok(val) => val,
+        Err(_) => {
+            return ProcMacroResult::new(TokenStream::empty())
+                .with_diagnostics(Diagnostic::error("Invalid base value").into());
+        }
+    };
+
+    let exp: usize = match macro_args[1].parse() {
+        Ok(val) => val,
+        Err(_) => {
+            return ProcMacroResult::new(TokenStream::empty())
+                .with_diagnostics(Diagnostic::error("Invalid exponent value").into());
+        }
+    };
+
+    let result: BigDecimal = pow(base, exp);
+
+    ProcMacroResult::new(TokenStream::new(result.to_string()))
+}
+#[inline_macro]
+pub fn log(token_stream: TokenStream) -> ProcMacroResult {
+    let db = SimpleParserDatabase::default();
+    let (parsed, _diag) = db.parse_virtual_with_diagnostics(token_stream);
+
+    let mut args = parsed.descendants(&db).filter(|node| 
+        matches!(node.kind(&db), 
+            SyntaxKind::Arg         )
+    );
+
+
+
+    // Extract log level
+    let level = match args.next() {
+        Some(node)  => node.get_text(&db),
+        _ => {
+            return ProcMacroResult::new(TokenStream::empty())
+            .with_diagnostics(Diagnostic::error("Invalid log level").into())
+        },
+    };
+
+    // Validate log level
+    println!("Level: {}", level);
+    if level != "\"TRACE\"" && level != "\"DEBUG\"" {
+        return ProcMacroResult::new(TokenStream::empty())
+            .with_diagnostics(Diagnostic::error("Invalid log level. Use TRACE or DEBUG").into());
+    }
+
+    // Extract format string
+    let format_string = match args.next() {
+        Some(node) => node.get_text(&db),
+        _ => return ProcMacroResult::new(TokenStream::empty())
+            .with_diagnostics(Diagnostic::error("Invalid format string").into()),
+    };
+
+    // Extract remaining arguments
+    let log_args: Vec<String> = args
+        .filter_map(|node| {
+            if matches!(node.kind(&db), SyntaxKind::Arg) {
+                Some(node.get_text(&db))
+            } else {
+                println!("Invalid argument: {:?} {:?}", node.get_text(&db), node.kind(&db));
+                None
+            }
+        })
+        .collect();
+
+       // Generate the log statement with feature checks
+       let log_statement = match level.as_str() {
+        "\"TRACE\"" => {
+            if log_args.is_empty() {
+                format!(
+                    "println!(\"{{}}: {{}}\", {}, format!({}))",
+                    "'TRACE'",
+                    format_string
+                )
+            } else {
+                format!(
+                    "println!(\"{{}}: {{}}\", {}, format!({}, {}))",
+                    "'TRACE'",
+                    format_string,
+                    log_args.join(", ")
+                )
+            }
+        },
+        "\"DEBUG\"" => {
+            if log_args.is_empty() {
+                format!(
+                    "println!(\"{{}}: {{}}\", {}, format!({}))",
```

Comment:
You should generate `println!` twice here, add `#[cfg(feature: 'log_level_trace')]` and `#[cfg(feature: 'log_level_debug')]` before each `println!`. This will generate the log when either `log_level_trace` or `log_level_debug` is enabled.

Reply:
#[test]
    fn test_log_macro(){
        #[cfg(feature: 'log_level_trace')]
        println!("hello");
    }

Unknown statement attribute.
 --> ../timestamp.cairo:144:9
        #[cfg(feature: 'log_level_trace')]
        

does this actually work in Cairo?, i tried this but can't compile. As I understand the `log!` will expand to something like this right?

Reply:
You are right, `cfg` can't be attached to statements.

Reply:
For now the best thing we can achieve is, in log! macro generate code that surrounds println! with if:
```
if LOG_LEVEL_TRACE {
  printlnn!(...)
}
```
and in log.cairo conditionally declare constants:
```
#[cfg(feature: 'log_level_trace')]
const LOG_LEVEL_TRACE = true;

#[cfg(feature: 'log_level_trace')]
const LOG_LEVEL_DEBUG = true;

#[cfg(feature: 'log_level_debug')]
const LOG_LEVEL_TRACE = false;

#[cfg(feature: 'log_level_debug')]
const LOG_LEVEL_DEBUG = true;

#[cfg(feature: 'log_level_none')]
const LOG_LEVEL_TRACE = false;

#[cfg(feature: 'log_level_none')]
const LOG_LEVEL_DEBUG = false;
```

`log_level_none` is enabled by default.

Reply:
For now the best thing we can achieve is, in log! macro generate code that surrounds println! with if:
```
if LOG_LEVEL_TRACE {
  printlnn!(...)
}
```
and in log.cairo conditionally declare constants:
```
#[cfg(feature: 'log_level_trace')]
const LOG_LEVEL_TRACE = true;

#[cfg(feature: 'log_level_trace')]
const LOG_LEVEL_DEBUG = true;

#[cfg(feature: 'log_level_debug')]
const LOG_LEVEL_TRACE = false;

#[cfg(feature: 'log_level_debug')]
const LOG_LEVEL_DEBUG = true;

#[cfg(feature: 'log_level_none')]
const LOG_LEVEL_TRACE = false;

#[cfg(feature: 'log_level_none')]
const LOG_LEVEL_DEBUG = false;
```

`log_level_none` is enabled by default.

Reply:
And, it seems, is not that bad:
 ![image](https://github.com/user-attachments/assets/964f220a-a929-4c14-89b5-9f829ba1ecff)